warehouse/manage/views.py

```diff
@@ -288,6 +326,61 @@ def delete_account(self):
         return logout(self.request)
 
 
+@view_defaults(
+    route_name="manage.account.totp-provision",
+    renderer="manage/account/totp-provision.html",
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+    permission="manage:user",
+)
+class ProvisionTOTPViews:
+    def __init__(self, request):
+        self.request = request
+        self.user_service = request.find_service(IUserService, context=None)
+
+    @view_config(request_method="GET")
+    def totp_provision(self):
+        if self.request.user.totp_provisioned:
+            self.request.session.flash("TOTP already provisioned.", queue="error")
+            return HTTPSeeOther(self.request.route_path("manage.account"))
+
+        totp_secret = self.request.user.totp_secret
+        if totp_secret is None:
+            totp_secret = otp.generate_totp_secret()
+            self.user_service.update_user(self.request.user.id, totp_secret=totp_secret)
+
+        totp_uri = self.user_service.totp_provisioning_uri(self.request.user.id)
+
+        return {
+            "provision_totp_form": ProvisionTOTPForm(user_service=self.user_service),
+            "provision_totp_uri": totp_uri,
+        }
+
+    @view_config(request_method="POST", request_param=ProvisionTOTPForm.__params__)
+    def validate_totp_provision(self):
+        if self.request.user.totp_provisioned:
+            self.request.session.flash("TOTP already provisioned.", queue="error")
+            return HTTPSeeOther(self.request.route_path("manage.account"))
+
+        form = ProvisionTOTPForm(**self.request.POST, user_service=self.user_service)
+        totp_uri = self.user_service.totp_provisioning_uri(self.request.user.id)
+
+        if form.validate():
+            totp_value = form.totp_value.data.encode("ascii")
```

Comment:
Just wondering why the `ascii` codec is explicitly used here rather than the default `utf-8`? (I'm assuming `form.totp_value.data` is a `str`). Does `form.validate()` guarantee that the data here will be ascii?

Reply:
No particular reason! `encode()` with the default UTF-8 encoding would suffice. I'll make this change tomorrow.

`form.validate()` does indeed ensure that the data is ASCII, by virtue of only accepting `/[0-9]{6}/`.